python/cuml/cuml/neighbors/nearest_neighbors.pyx

```diff
@@ -674,22 +674,82 @@ class NearestNeighbors(UniversalBase,
                 ).array
                 I_ndarr.index = index
 
+        if use_training_data:
+            I_ndarr = I_ndarr.to_output('cupy')
+            D_ndarr = D_ndarr.to_output('cupy')
+
+            rows, cols = I_ndarr.shape
+
+            # Kernel to check for zeros in the second column of D_ndarr
+            check_zero_kernel = cp.RawKernel(r'''
```

Comment:
cupy will cache these based on the source, so the actual compiler won't be invoked more than once. However it does do quite a few steps _before_ the cache check is hit. Could you instead move these kernel definitions to be module level? They won't be compiled until first use, but this will reduce the overhead to basically nothing after that.